### PR TITLE
level_1/task_7/part_3

### PR DESCRIPTION
**Pull request by PolinaXox**
html, css, js and php codes are in different files.
Between 'less code and use more server's globals' and 'more code but all needed request parts are checked' I've choosen the first one. But it's too interesting to try the second option, so I'll try.

### CODE REVIEW

File: 1. Basics/task_7/part_3/scripts/task_7_part_3.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * Changes form`s method
+ * 
+ * @param {*} elementID 
+ */
+function clickOnCheckBoxEsRequestMethodIsNotGET(elementID) {
+	let form = document.getElementById("myForm");
+	let checkBox = document.getElementById(elementID);
+	
+	if(checkBox.checked == true) {
+		form.method = 'POST';
```
[the remark this answers is not in the document]
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/scripts/task_7_part_3.js]: Done.

File: 1. Basics/task_7/part_3/scripts/task_7_part_3.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * 
+ * @param {*} elementID 
+ */
+
+function clickOnCheckBoxEsRequestWrongUriStart(elementID) {
+	let form = document.getElementById("myForm");
+	let checkBox = document.getElementById(elementID);
+	
+	if (checkBox.checked == true) {
+    	myForm.action = "/notsum";
```
[the remark this answers is not in the document]
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/scripts/task_7_part_3.js]: Done.

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+/**
+ * @param string $method
+ * @param string $uri
+ * @param string|bool $nums
+ * 
+ * @throws Exception
+ * 
+ * @return int
+ */
+function getResponseStatusCode(string $method, string $uri, string|bool $nums) : int {
```
[the remark this answers is not in the document]
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/task_3_on_apache_handler.php]: Now it's name is `checkRequestCorrectness()` and the function was changed corresponding to the new approach: less attention to the headers.

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+ * 
+ * @param string $requestMethod
+ * @param string $requestUri
+ * @param string|bool $nums
+ * 
+ * @return array
+ */
+function getResponseDraft(string $requestMethod, string $requestUri, string|bool $nums) : array {
+    try {
+        $statusCode = getResponseStatusCode($requestMethod, $requestUri, $nums);
```
[the remark this answers is not in the document]
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/task_3_on_apache_handler.php]: The new approach "Less attention to the headers" discarded this function.

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+/**
+ * Changes input data to prevent harm, injections ect.
+ * (Able to use as a callback function for array_walk() and for array_walk_recursive())
+ * 
+ * @param string $data
+ * 
+ * @return void
+ */
+ function testInputValueByReference(string &$data) : void { //
+	$data = trim($data);
```
Comment (anton-moroz) [inline 1. Basics/task_7/part_3/task_3_on_apache_handler.php:157]: and where is test?

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+ * Checks the request's correctness.
+ * If the request is correct return false.
+ * If the request is incorrect returns corresponding response code and some additional information.
+ *
+ * @param string|bool $nums
+ * @return bool|array
+ */
+function checkRequestCorrectness(string|bool $nums) : bool|array {
+	$methodGet = (($_SERVER['REQUEST_METHOD'] ?? null) === 'GET') ? true : false;				// 400
+	$correctUri = preg_match('#^/sum#', $_SERVER['REQUEST_URI'] ?? '');		// 404
```
Comment (anton-moroz) [inline 1. Basics/task_7/part_3/task_3_on_apache_handler.php:47]: this way is not most optimal.
In any case you will do a lot of checks, even if not all of them are needed. For example, if method is invalid you anyway check uri and numbers...
no sense in this and you waste time and resources

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+/**
+ * Processes http request, partly forms the response. 
+ * Optionally outputs the results as HTML page.
+ * 
+ * @return void
+ */
+function processHttpRequest() : void {
+	$nums = testInputValue($_GET['nums'] ?? false);
+	$invalidRequest = checkRequestCorrectness($nums);
+	$isHTML = preg_match('#html#', $_SERVER['HTTP_ACCEPT'] ?? '');
```
Comment (anton-moroz) [inline 1. Basics/task_7/part_3/task_3_on_apache_handler.php:23]: same as below... this check in needed later, but not here

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php (hunk 5 above)
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/task_3_on_apache_handler.php:157]: OK. `makeInputDataSafe...()` sounds better.

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php (hunk 6 above)
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/task_3_on_apache_handler.php:47]: Caught it. Done.

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php (hunk 7 above)
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/task_3_on_apache_handler.php:23]: Done.

**Review by anton-moroz — CHANGES_REQUESTED** [review]

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+/**
+ * Changes input data to prevent harm, injections ect.
+ * 
+ * @param string $data
+ * 
+ * @return string
+ */
+ function makeInputDataSafe(string $data) : string {
+	makeInputDataSafeByReference($data);
+	return $data;
```
[the remark this answers is not in the document]
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/task_3_on_apache_handler.php:180]: Done.

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+	// 404
+	$correctUri = preg_match('#^/sum#', $_SERVER['REQUEST_URI'] ?? '');
+	if(!$correctUri) return ['code' => 404, 'info' => 'Not found. URI is invalid.'];
+
+	// 400
+	if(!$nums) return ['code' => 400, 'info' => 'Value \'nums\' is absent.'];
+
+	// 400
+	$correctNums = preg_match('#^(\d+,?)+$#', $nums);
```
[the remark this answers is not in the document]
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/task_3_on_apache_handler.php]: Done.

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+ */
+function checkRequestCorrectness(string|bool $nums) : bool|array {
+
+	// 400
+	$methodGet = (($_SERVER['REQUEST_METHOD'] ?? null) === 'GET') ? true : false;
+	if(!$methodGet) return ['code' => 400, 'info' => 'Request method is invalid. The method GET is needed.'];
+
+	// 404
+	$correctUri = preg_match('#^/sum#', $_SERVER['REQUEST_URI'] ?? '');
+	if(!$correctUri) return ['code' => 404, 'info' => 'Not found. URI is invalid.'];
```
[the remark this answers is not in the document]
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/task_3_on_apache_handler.php]: Done

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+ * If the request is incorrect returns corresponding response code and some additional information.
+ *
+ * @param string|bool $nums
+ * @return bool|array
+ */
+function checkRequestCorrectness(string|bool $nums) : bool|array {
+
+	// 400
+	$methodGet = (($_SERVER['REQUEST_METHOD'] ?? null) === 'GET') ? true : false;
+	if(!$methodGet) return ['code' => 400, 'info' => 'Request method is invalid. The method GET is needed.'];
```
[the remark this answers is not in the document]
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/task_3_on_apache_handler.php]: Done. Now it works without variable.

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+// CHECK INPUT DATA FUNCTIONS //////////////////////////////////////////////////
+/**
+ * Changes input data to prevent harm, injections ect.
+ * (Able to use as a callback function for array_walk() and for array_walk_recursive())
+ * 
+ * @param string $data
+ * 
+ * @return void
+ */
+ function makeInputDataSafeByReference(string &$data) : void { //
```
Comment (anton-moroz) [inline 1. Basics/task_7/part_3/task_3_on_apache_handler.php:164]: name this function `sanitizeInput`

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+/**
+ * Changes input data to prevent harm, injections ect.
+ * 
+ * @param string $data
+ * 
+ * @return string
+ */
+ function makeInputDataSafe(string $data) : string {
```
Comment (anton-moroz) [inline 1. Basics/task_7/part_3/task_3_on_apache_handler.php:177]: name this function 'sanitizeInput'

**Review by anton-moroz — CHANGES_REQUESTED** [review]

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php (hunk 12 above)
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/task_3_on_apache_handler.php]: Quite a nice name. Thank you!

File: 1. Basics/task_7/part_3/task_3_on_apache_handler.php (hunk 13 above)
Comment (PolinaXox) [reply 1. Basics/task_7/part_3/task_3_on_apache_handler.php]: Done

**Pull request merged**